client-spark/common/src/main/java/org/apache/spark/shuffle/handle/ShuffleHandleInfo.java

```diff
@@ -29,6 +29,13 @@ public interface ShuffleHandleInfo {
   /** Get all the assigned servers including the excluded servers. */
   Set<ShuffleServerInfo> getServers();
 
+  /**
+   * When a Stage retry occurs, replace the current PartitionToShuffleServer and record the
+   * historical PartitionToShuffleServe.
+   */
+  default void replaceCurrentShuffleHandleInfo(
```

Comment:
I think there is no need to introduce the extra general interface method here. If you want to update the handleInfo, you could forcelly transform type into ChainShuffleHandleInfo and then to update its inner current handle.